# Computing with formulas
*AKA I bought an expensive computer and I use it as if were a cheap calculator*

## A formula

Evaluate $y(t) = v_0t - \frac{1}{2}gt^2$ for $v_0=0.5 [\text{m} \text{s}^{-1}]$, $g=9.81 [\text{m} \text{s}^{-1}]$ and $t=0.6 [\text{s}]$

In [1]:
print 5*0.6 - 0.5*9.81*0.6**2

1.2342


Evaluate again $y(t) = v_0t - \frac{1}{2}gt^2$, now for $v_0=1 [\text{m} \text{s}^{-1}]$ and $t=0.1 [\text{s}]$, while keeping the same acceleration $g=9.81 [\text{m} \text{s}^{-1}]$

In [2]:
print 1*0.1 - 0.5*9.81*0.1**2

0.05095


### Variables

Other ways of evaluating $y(t) = v_0t - \frac{1}{2}gt^2$ for $v_0=0.5 [\text{m} \text{s}^{-1}]$, $g=9.81 \text{m} \text{s}^{-1}$ and $t=0.6 \text{s}$ 

We now define **variables** to denote relevant quantities ('initial velocity', the 'acceleraton of gravity' and 'time') assign them a numerical value (or an expression of already defined variables)

In [3]:
v0 = 5
g = 9.81
t = 0.6

y = v0*t - 0.5*g*t**2 

print y

1.2342


**Variable names** can contain any lower or upper case letter or numbers from 0 to 9 (the first character of the name cannot be a number though)

* Long variables names are explicative but can make the code visually cluttered

In [4]:
initial_velocity = 5
acceleration_of_gravity = 9.81
TIME = 0.6

VerticalPositionOfBall = initial_velocity*TIME - 0.5*acceleration_of_gravity*TIME**2

print VerticalPositionOfBall

1.2342


Whenever a Python variable is used to denote a mathematical symbol (for instance, an actual mathematical variable in an actual mathematical formula), it is good practice to name in a way that it resembles the original symbol

### Comments

Python **comments** are important to maintain an intuitive connection between mathematical expressions and the code used to represent/evaluate them

In [5]:
# Compute the height of a ball in vertical motion

v0 = 5                # initial velocity
g = 9.81              # acceleration of gravity
t = 0.6               # time

y = v0*t - 0.5*g*t**2 # vertical position

print y

1.2342


### Prettier printing

Printing the results should be done in such a way that what gets outputted is informative

#### The printf syntax for formatting

In [6]:
print 'At t=%g s, the height of the ball is %.2f m.' % (t, y)

At t=0.6 s, the height of the ball is 1.23 m.


The string has '*slots*' starting with a percentage sign (here, '**%g**' and '**%.2f**') in which Python variables can be inserted

* **The percent sign marks the slot**
* **The following character sets the format specification**

The variables are listed, separated by commas, inside parenthesis after the string (here, '**% (t, y)**')

* **The first variable goes into the first slot**
* **The second variable goes into the second slot**



##### An example

Various format specifications

In [7]:
i = 62
r = 189876545.7654675432

# Print out numbers with quotes "" to see width of field

print '"%d"' % i       # minimum field
print '"%5d"' % i      # field of width 5 characters
print '"%05d"' % i     # pad with zeros

print '"%g"' % r       # r is big number, scientific notation
print '"%G"' % r       # E in the exponent
print '"%e"' % r       # compact scientific notation
print '"%E"' % r       # compact scientific notation
print '"%20.2E"' % r   # 2 decimals, field of width 20
print '"%30g"' % r     # field of width 30 (right-adjusted)
print '"%-30g"' % r    # left-adjust number
print '"%-30.4g"' % r  # 3 decimals

print '%s' % i         # convert i to string automatically
print '%s' % r

# Use %% to print the percentage sign
print '%g %% of %.2f Euro is %.2f Euro' % (5.1, 346, 5.1/100*346)


"62"
"   62"
"00062"
"1.89877e+08"
"1.89877E+08"
"1.898765e+08"
"1.898765E+08"
"            1.90E+08"
"                   1.89877e+08"
"1.89877e+08                   "
"1.899e+08                     "
62
189876545.765
5.1 % of 346.00 Euro is 17.65 Euro


##### An example

Triple-quoted strings for multi-line printing

In [8]:
v0 = 5
g = 9.81
t = 0.6

y = v0*t - 0.5*g*t**2

print """
At t=%f s, a ball with
initial velocity v0=%.3E m/s
is located at the height %.2f m.
""" % (t, v0, y)


At t=0.600000 s, a ball with
initial velocity v0=5.000E+00 m/s
is located at the height 1.23 m.



#### The format syntax for formatting

An alternative to **printf**, with more functionalities and a different syntax

* Slots are denoted by curly braces (rather than a percentage sign)
* Variable are listed with an optional colon and format specifier
* Variables and their values are listed at the end of the statement
* Slots have names (the sequence of variables is not important)

In [9]:
print 'At t={t:g} s, the height of the ball is {y:.2f} m.' .format(t=t, y=y)

print 'At t={t:E} s, the height of the ball is {y:.4g} m.' .format(t=t*t/t, y=y+2*t-t*2)

At t=0.6 s, the height of the ball is 1.23 m.
At t=6.000000E-01 s, the height of the ball is 1.234 m.


In [10]:
print """
At t={t:f} s, a ball with
initial velocity v0={v0:.3E} m/s
is located at the height {y:.2f} m.
""".format(t=t, v0=v0, y=y)


At t=0.600000 s, a ball with
initial velocity v0=5.000E+00 m/s
is located at the height 1.23 m.



## Another formula

Evaluate $F = \cfrac{9}{5}C + 32$ for, say, $C=21C$

In [11]:
C = 21
F = (9/5)*C + 32
print F

53


$53$ is not the correct answer, as $\cfrac{9}{5}21 + 32 = 1.8 \cdot 21 + 32 = 37.8 + 32 = 69.8$

The problem is in the evaluation of $\cfrac{9}{5}$

### Integer division and float division

In many computer languages there are two type of divisions

* ** Float division**
* ** Integer division**

##### Float division

$9/5 = 1.8$, in decimal notation

##### Integer division

$9/5 = 1$

Let $a$ and $b$ be two integers, $a/b$ is also an integer
* It is the largest integer $c$ such that $bc \le a$

To avoid accidental integer division:
* Double forward-slash (**//**)
* Argument to interpeter (**-Qnew** and/or **-Qwarnall**)
* Divisions imported from future (**from \_\_future\_\_ import division**)

### Objects

* **Integer objects**
* **Float objects**
* **String objects**

## Mathematical functions

### An example

Solve $\cfrac{1}{2}gt^2 - v_0t + y_c = 0$ to get 
\begin{cases} 
t_1 = \cfrac{v_0 - \sqrt{v_0^2 - 2gy_c}}{g} \\
t_1 = \cfrac{v_0 + \sqrt{v_0^2 - 2gy_c}}{g}
\end{cases}

We need the 'square root' function from the 'math module'
* We must import the module to make its functions available
* **import module\_name** then **module\_name.function\_name** to use

In [12]:
v0 = 5
g = 9.81
yc = 0.2

import math
t1 = (v0 - math.sqrt(v0**2 - 2*g*yc))/g
t2 = (v0 + math.sqrt(v0**2 - 2*g*yc))/g

print 'At t=%g s and %g s, the height is %g m.' % (t1, t2, yc)

At t=0.0417064 s and 0.977662 s, the height is 0.2 m.


Alternatively, we can use an alternative import syntax
* **from module\_name import function\_name** and then **funtion_name** to use

In [13]:
from math import sqrt

t1 = (v0 - sqrt(v0**2 - 2*g*yc))/g
t2 = (v0 + sqrt(v0**2 - 2*g*yc))/g

All functions in a module can be imported at once
* **from module\_name import \* **
* **import module\_name as short\_module\_name**

In [14]:
import math as m      # m is now the name of the math module

v = m.sin(m.pi)
print v

from math import log as ln
v = ln(5)
print v

from math import sin as s, cos as c, log as ln
v = s(5)*c(5) + ln(5)
print v

1.22464679915e-16
1.60943791243
1.33742735699


### An example

Consider function $\sinh{x} = \cfrac{1}{2}\Big(e^x - e^{-x}\Big)$ and evaluate it for some value of $x$

* Using **sinh.math** directly
* Using **math.exp**
* Using **math.e**

In [15]:

from math import sinh, exp, e, pi

x = 2*pi

r1 = sinh(x)
r2 = 0.5*(exp(x) - exp(-x))
r3 = 0.5*(e**x - e**(-x))

print r1
print r2
print r3

267.744894041
267.744894041
267.744894041


The valuations apparently led to identical results, but because of **rounding errors** they aren't 
* This is easily shown by using a more exhaustive format for displaying the results

In [16]:
print '%.16f' % (r1)
print '%.16f' % (r2)
print '%.16f' % (r3)

267.7448940410164369
267.7448940410164369
267.7448940410163232


Even more trivial calculations are prone to rounding errors
* $\cfrac{1}{49}49$ 
* $\cfrac{1}{51}51$

In [17]:
print '%.16f %.16f' % (1/49.0*49, 1/51.0*51)

0.9999999999999999 1.0000000000000000


### Type conversion

In [18]:
print '--- Case A ---'
C=21; print C; print type(C)
print '--- ------ --- \n'

print '--- Case B ---'
C = float(C); print C; print type(C)
print '--- ------ --- \n'

print '--- Case C ---'
C = 20.9; print C; print type(C)
print '--- ------ --- \n'

print '--- Case D ---'
D = int(C); print D; print type(D)
print '--- ------ --- \n'



--- Case A ---
21
<type 'int'>
--- ------ --- 

--- Case B ---
21.0
<type 'float'>
--- ------ --- 

--- Case C ---
20.9
<type 'float'>
--- ------ --- 

--- Case D ---
20
<type 'int'>
--- ------ --- 



In [19]:
print int(20.9)        # Truncation

print round(20.9)      # Rounding

print int(round(20.9)) # Type conversion after rounding

20
21.0
21


## Complex numbers

### Complex arithmetics

In [20]:
u = 2.5 + 3j      # create a complex number
v = 2             # this is an integer
w = u+v           # complex + integer
print w

a = -2            #
b = 0.5           #
s = a+b*1j       # complex number from two floats
print s

import cmath
    
s2 = complex(a,b) # alternative way
print s2

print s*w               # complex*complex
print s/w               # complex/complex

(4.5+3j)
(-2+0.5j)
(-2+0.5j)
(-10.5-3.75j)
(-0.25641025641+0.282051282051j)


A **complex object** is equipped with the usual functionalities

In [21]:
s = -2+0.5j; print 'A complex number: \n', s, '\n'

print 'Its real part: ', s.real
print 'Its imaginary part: ', s.imag

print '\n', 'The complex conjugate: \n', s.conjugate()

A complex number: 
(-2+0.5j) 

Its real part:  -2.0
Its imaginary part:  0.5

The complex conjugate: 
(-2-0.5j)


### Complex functions

The **cmath module** has functions that return complex numbers (objects)

**NumPy** has mathematical functions that return a float object or a complex object when appropriate
* **from numpy.lib.scimath import **
* **from scipy import **
* **from scitools.std import **

In [22]:
## MATH
print '--- MATH ---'
from math import sqrt
print sqrt(4); print type(sqrt(4))                  # Float
#print sqrt(-1)                                     # Illegal (uncomment to run)
print '--- ---- --- \n'


## CMATH
print '--- CMATH ---'
from cmath import sqrt
print sqrt(4);  print type(sqrt(4))                 # Complex
print sqrt(-1); print type(sqrt(-1))                # Complex
print '--- ----- --- \n'

## NUMPY/SCIMATH
print '--- NUMPY/SCIMATH ---'
from numpy.lib.scimath import *
print sqrt(4);  print type(sqrt(4))                 # Float
print sqrt(-1); print type(sqrt(-1))                #Complex
print '--- ------------- ---'

--- MATH ---
2.0
<type 'float'>
--- ---- --- 

--- CMATH ---
(2+0j)
<type 'complex'>
1j
<type 'complex'>
--- ----- --- 

--- NUMPY/SCIMATH ---
2.0
<type 'numpy.float64'>
1j
<type 'numpy.complex128'>
--- ------------- ---


Consider function $f(x) = ax^2 + bx + c$

In [23]:
a = 1; b = 2; c = 100             # polynomial coefficients

from numpy.lib.scimath import sqrt

r1 = (-b + sqrt(b**2 - 4*a*c))/(2*a)
r2 = (-b - sqrt(b**2 - 4*a*c))/(2*a)

print r1
print r2

(-1+9.94987437107j)
(-1-9.94987437107j)


In [24]:
from math import sin
#r = sin(w)         # Uncomment to run

In [25]:
from cmath import sin, sinh

r1 = sin(8j)
print r1

r2 = 1j*sinh(8)
print r2

1490.47882579j
1490.47882579j


Another known relations is $e^{iq} = \cos{(q)} + i\sin{(q)}$

In [26]:
from cmath import cos, exp

q = 8           

print exp(1j*q)
print cos(q) + 1j*sin(q)

(-0.145500033809+0.989358246623j)
(-0.145500033809+0.989358246623j)


## Symbolic mathematics

Package **SymPy** allows performing symbolic operations, like differentiation and integration, equations solving, series expansions and so on

 * **symbols**, define symbols for symbolic math
 * **diff**, differentiate expressions
 * **integrate**, integrate expressions
 * **Rational**, define rational numbers
 * **lambdify**, turn symbolic expressions into Python functions

### Differentiation and integration

Consider function $y=v_0t - \cfrac{1}{2}g^2$ and its derivative with respect to the independent variable $t$

In [27]:
from sympy import (symbols,diff,integrate,Rational,lambdify)

t, v0, g = symbols('t v0 g')

y = v0*t - Rational(1,2)*g*t**2; print y  # Function (position)
dydt = diff(y,t); print dydt              # First derivative (velocity)

print 'acceleration:', diff(y,t,t)        # Second derivative (acceleration)

y2 = integrate(dydt, t); print y2         # Integral of the first derivative (function/position)

-g*t**2/2 + t*v0
-g*t + v0
acceleration: -g
-g*t**2/2 + t*v0


SymPy allows symbolic expressions to be turned into ordinary numerical functions

In [28]:
v = lambdify([t, v0, g], dydt); # Arguments in v [t, v0, g] and symbolic expression dydt
print type(v)

print v(2,5,9.81) # Calculation
print 5 - 9.81*2  # Manual check


<type 'function'>
-14.62
-14.62


### Equation solving

Consider $y = v_0t - \cfrac{1}{2}gt^2$ and symbolically solve for $t$ when $y=0$

In [29]:
from sympy import solve
roots = solve(y,t); print roots

[0, 2*v0/g]


In [30]:
print y.subs(t, roots[0])
print y.subs(t, roots[1])

0
0


### Taylor series

Consider functions $e^t$ and $e^{\sin{(t)}}$ in the independent variable $t$ and calculate their Taylor's expansion around point $t_0$

In [31]:
from sympy import exp, sin, cos

f = exp(t); print 'Third-order Taylor expansion of function exp(t) around t=0: \n', f.series(t,0,3), '\n'
f = exp(sin(t)); print 'Eighth-order Taylor expansion of function sin(x) around t=0: \n', f.series(t,0,8)

Third-order Taylor expansion of function exp(t) around t=0: 
1 + t + t**2/2 + O(t**3) 

Eighth-order Taylor expansion of function sin(x) around t=0: 
1 + t + t**2/2 - t**4/8 - t**5/15 - t**6/240 + t**7/90 + O(t**8)


It is possible to output the results using LaTeX syntax

In [32]:
from sympy import latex

print 'Corresponding LaTeX sybtax: \n', latex(f.series(t,0,7))

Corresponding LaTeX sybtax: 
1 + t + \frac{t^{2}}{2} - \frac{t^{4}}{8} - \frac{t^{5}}{15} - \frac{t^{6}}{240} + \mathcal{O}\left(t^{7}\right)


That is, $1 + t + \frac{t^{2}}{2} - \frac{t^{4}}{8} - \frac{t^{5}}{15} - \frac{t^{6}}{240} + \mathcal{O}\left(t^{7}\right)$

There are also toosl for expand and simplify expressions

In [33]:
from sympy import simplify, expand

x, y = symbols('x y')
f = -sin(x)*sin(y) + cos(x)*cos(y); print 'Function f: ', f

print 'Simplified function f: ', simplify(f), '\n'                   # Known trigonometric identity
print 'Expanded function sin(x+y): ', expand(sin(x + y), trig=True)  # Needs trigonometric hint

Function f:  -sin(x)*sin(y) + cos(x)*cos(y)
Simplified function f:  cos(x + y) 

Expanded function sin(x+y):  sin(x)*cos(y) + sin(y)*cos(x)
